### PR TITLE
OnnxPrecomputedHybridSystem: recog at best epoch

### PR DESCRIPTION
**Pull request by vieting**
This PR adds support to allow specifying `"best"` as epoch for the recognition using the `OnnxPrecomputedHybridSystem`. It will use the `GetBestPtCheckpointJob` and the resulting output checkpoint.

### CODE REVIEW

File: common/setups/rasr/onnx_precomputed_hybrid_system.py [hunk 1]
```diff
@@ -56,15 +57,25 @@ def nn_recognition(
             epochs = epochs if epochs is not None else list(checkpoints.keys())
 
             for pron, lm, prior, epoch in itertools.product(pronunciation_scales, lm_scales, prior_scales, epochs):
-                assert epoch in checkpoints.keys()
+                if epoch == "best":
+                    assert train_job is not None, "train_job needed to get best epoch checkpoint"
+                    best_checkpoint_job = GetBestPtCheckpointJob(
+                        train_job.out_model_dir, train_job.out_learning_rates, key="dev_loss_CE", index=0
```
Comment (Atticus1806) [inline common/setups/rasr/onnx_precomputed_hybrid_system.py:63]: I know this might get messy, but can we always assume "dev_loss_CE" to exist like this here?

Comment (vieting) [reply common/setups/rasr/onnx_precomputed_hybrid_system.py:63]: Good point. Off the top of my head, I don't see a very clean and generic solution here. Any suggestions?

File: common/setups/rasr/onnx_precomputed_hybrid_system.py [hunk 2]
```diff
@@ -36,6 +36,7 @@ def nn_recognition(
         mem: int,
         nn_prior: bool,
         epochs: Optional[List[int]] = None,
```
[the remark this answers is not in the document]
Comment (vieting) [reply common/setups/rasr/onnx_precomputed_hybrid_system.py:38]: Right.
```suggestion
        epochs: Optional[List[int, str]] = None,
```

File: common/setups/rasr/onnx_precomputed_hybrid_system.py (hunk 1 above)
Comment (JackTemaki) [reply common/setups/rasr/onnx_precomputed_hybrid_system.py:63]: It just has to be configurable, and I am fine having this as default

File: common/setups/rasr/onnx_precomputed_hybrid_system.py [hunk 3]
```diff
@@ -35,7 +35,8 @@ def nn_recognition(
         rtf: int,
         mem: int,
         nn_prior: bool,
-        epochs: Optional[List[int]] = None,
+        epochs: Optional[List[int, str]] = None,
+        train_job: Optional[Union[returnn.ReturnnTrainingJob, returnn.ReturnnRasrTrainingJob]] = None,
         needs_features_size: bool = True,
         acoustic_mixture_path: Optional[tk.Path] = None,
```
Comment (vieting) [inline common/setups/rasr/onnx_precomputed_hybrid_system.py:41]:
```suggestion
        acoustic_mixture_path: Optional[tk.Path] = None,
        best_checkpoint_key: str = "dev_loss_CE",
```

File: common/setups/rasr/onnx_precomputed_hybrid_system.py (hunk 1 above)
Comment (vieting) [reply common/setups/rasr/onnx_precomputed_hybrid_system.py:63]:
```suggestion
                        train_job.out_model_dir, train_job.out_learning_rates, key=best_checkpoint_key, index=0
```

**Review by Atticus1806 — APPROVED** [review]

**Review by JackTemaki — COMMENTED** [review]
I am not sure how happy I am with passing around whole jobs, but I guess creating an extra structure for passing all paths and the learning rate file is maybe overkill...

**Comment by vieting** [comment]
> I am not sure how happy I am with passing around whole jobs, but I guess creating an extra structure for passing all paths and the learning rate file is maybe overkill...

Creating an extra structure also seems like overkill to me. What speaks against passing the job in general?

**Comment by vieting** [comment]
@JackTemaki?

**Review by JackTemaki — APPROVED** [review]
Sorry!
